src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -1043,6 +1049,97 @@ private PackageLicenseInfo GeneratePackageLicenseInfo(IPackageSearchMetadata met
+        private Dictionary<string, List<string>>? GetSourceMappingsForPackages(IEnumerable<PackageIdentity> packages, PackageSourceMapping packageSourceMapping)
+        {
+            Dictionary<string, List<string>> packageIdToSourceNames = new();
+            foreach (PackageIdentity package in packages)
+            {
+                string packageId = package.Id;
+                IReadOnlyList<string>? sourcesMappedToPackage = packageSourceMapping.GetConfiguredPackageSources(packageId);
+                if (sourcesMappedToPackage is null)
+                {
+                    return null;
+                }
+
+                foreach (string sourceName in sourcesMappedToPackage)
+                {
+                    if (!packageIdToSourceNames.ContainsKey(packageId))
+                    {
+                        packageIdToSourceNames.Add(packageId, new List<string>());
+                    }
+                    packageIdToSourceNames[packageId].Add(sourceName);
+                }
+            }
+
+            return packageIdToSourceNames;
+        }
+
+        private async Task<IPackageSearchMetadata[]> GetOnlyLocalPackageMetadataAsync(
+            INuGetUI uiService,
+            SourceCacheContext sourceCacheContext,
+            IEnumerable<PackageIdentity> packages,
+            CancellationToken token)
+        {
+            var projects = (IReadOnlyCollection<IProjectContextInfo>)uiService.Projects;
+            var searchService = uiService.UIContext.ReconnectingSearchService;
+            IReadOnlyList<SourceRepository> localSources = await searchService.GetAllPackageFoldersAsync(projects, token);
+
+            IPackageSearchMetadata[] completed = await GetPackageMetadataThrottledAsync(localSources, sourceCacheContext, packages, token);
+            return completed;
+        }
+
+        private static async Task<IPackageSearchMetadata[]> GetPackageMetadataThrottledAsync(IEnumerable<SourceRepository> sources, SourceCacheContext sourceCacheContext, IEnumerable<PackageIdentity> packages, CancellationToken token)
+        {
+            IPackageSearchMetadata[] completed = (await TaskCombinators.ThrottledAsync(
+                packages,
+                (p, t) => GetPackageMetadataAsync(sources, sourceCacheContext, p, t),
+                token)).Where(metadata => metadata != null).Cast<IPackageSearchMetadata>().ToArray();
+
+            return completed;
+        }
+
+        private static async Task<IPackageSearchMetadata[]> GetPackageMetadataThrottledAsync(PackageSourceMapping packageSourceMapping, IEnumerable<SourceRepository> enabledSources, SourceCacheContext sourceCacheContext, IEnumerable<PackageIdentity> packages, CancellationToken token)
+        {
+            IPackageSearchMetadata[] completed = (await TaskCombinators.ThrottledAsync(
+                packages,
+                (p, t) =>
+                {
+                    var mappedSources = packageSourceMapping.GetConfiguredPackageSources(p.Id);
```

Comment:
Having the throttled delegate do the mapping lookup seemed like a better idea than precalculating the sources. Need to verify this works well.